Add stats of client and pool to be accessible through agent

tdeekens opened the pull request:
<!--
Before submitting a Pull Request, please read our contribution guidelines, which
can be found at CONTRIBUTING.md in the repository root.

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that tests and linting pass.

You will also need to ensure that your contribution complies with the
Developer's Certificate of Origin, outlined in CONTRIBUTING.md
-->

## This relates to:

This relates to the issue https://github.com/nodejs/undici/issues/4149

## Rationale

Integrating existing and exposed Pool stats into a telemetry system is a bit cumbersome. Right now we can hook into the `factory` function and observe the created pool there. It would be beneficial if a Pool's and Client's stats would be exposed by the Agent owning the Pool or Client respectively. 

## Changes

Exposing stats through the Agent makes it easier to integrate telemetry systems as we do not have to "hook" into the `factory` to receive the Pool or Client used by an Agent.

### Features

Exposes a new getter called `stats` on an Agent. The method returns an Array of Arrays with stats by origin. 

Additionally, the Client now also exposes available stats similar to a Pool. The initial implementation of Pool stats happened in [this PR](https://github.com/nodejs/undici/pull/1573) while the implementation for a Client here follows the Pool's implementation pattenr.

### Bug Fixes

<!-- List the fixed bugs here (if applicable), or write N/A if not -->

### Breaking Changes and Deprecations

<!-- List the breaking changes (changes that modify the existing API) and
deprecations (removed features) here -->

## Status

<!-- KEY: S = Skipped, x = complete -->


- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] Benchmarked (**optional**)
- [x] Documented
- [x] Review ready
- [x] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md#developers-certificate-of-origin

tdeekens commented on lib/core/symbols.js:
```diff
@@ -64,5 +64,6 @@ module.exports = {
   kHttpProxyAgent: Symbol('http proxy agent'),
-  kHttpsProxyAgent: Symbol('https proxy agent')
+  kHttpsProxyAgent: Symbol('https proxy agent'),
+  kStats: Symbol('stats')
```
This is used now by Pool and Client. Hence it should be shared in my opinion.

tdeekens commented on lib/dispatcher/agent.js:
```diff
@@ -110,6 +110,17 @@ class Agent extends DispatcherBase {
     await Promise.all(destroyPromises)
   }
+
+  get stats () {
```
Getter to loop over Clients which can be Pool or Client themselves to assemble and expose stats.

Data structure can be changed. Right now it's `[string, [PoolStats, ClientStats]]`. It could also be `Record<string, PoolStats | ClientStats>`.

metcoder95 replied:
The current shape seems like a good first iteration, especially for Agent as the number of origins called can vary; caller can reconstruct it into a different data structure if needed

tdeekens replied:
Thanks. Same thinking on my end.

tdeekens commented on lib/dispatcher/client-stats.js:
```diff
@@ -0,0 +1,28 @@
+'use strict'
```
Follows implementation of PoolStats. Clients however expose a subset of stats. Hence the separate class.

metcoder95 replied:
I'll put this class in `utils` or so, as the files within the `dispatcher/` dir are mostly mean for classes that shares `dispatcher` interface.

tdeekens replied:
Sounds good. Same for the PoolStats then which prior to this change lived here too, right?

tdeekens commented on lib/dispatcher/client.js:
```diff
@@ -249,6 +251,8 @@ class Client extends DispatcherBase {
     this[kResume] = (sync) => resume(this, sync)
     this[kOnError] = (err) => onError(this, err)
+
+    this[kStats] = new ClientStats(this)
```
Same as with Pools.

metcoder95 reviewed (commented):
Can we add testing for it?

tdeekens commented:
> Can we add testing for it?

Yip. I followed the Pool Stats addition PR in terms of depth of testing on only existing through tsd.

tdeekens commented:
Hi @metcoder95!

In [ab942e4](https://github.com/nodejs/undici/pull/4157/commits/ab942e49f42858f639ac6ab064deb05b1ac993da) I've moved the pool stats both of pool and client into a `utils/stats.js`. Happy for feedback. 
Furthermore in [f2452c4](https://github.com/nodejs/undici/pull/4157/commits/f2452c4f505f02e698253e9954bbb9f74e4c3312) I added some tests. Let me know if the direction is right and if the depth needs to be expanded on.

tdeekens commented:
👋🏼 @metcoder95 in [15bc63b](https://github.com/nodejs/undici/pull/4157/commits/15bc63b63f3437a938505be6999f6b91f1503e3a) I started adding some documentation where I thought it would help.

metcoder95 commented on mise.toml:
what is this exactly?

tdeekens replied:
Apologies. Will remove.

tdeekens replied:
[5a93529](https://github.com/nodejs/undici/pull/4157/commits/5a935298e870431d9879b0f5ae6950c241295ccd)

tdeekens commented:
Anything else you see should be improved @metcoder95? Thanks!

tdeekens commented:
@metcoder95 can you approve the workflows again. There was a type issue on the `connected` stat of a client which is a `boolean` compared to a pool where it is a `number`. Thanks!

tdeekens commented:
@metcoder95 I can't see that tests fail due to my changes with

```txt
Error: Cannot find module '../../deps/undici/src/index.js'
Require stack:
- /home/runner/work/undici/undici/node-v23.11.0/test/parallel/test-inspector-network-fetch.js
    at Function._resolveFilename (node:internal/modules/cjs/loader:1405:15)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1061:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1066:22)
    at Function._load (node:internal/modules/cjs/loader:1215:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Module.require (node:internal/modules/cjs/loader:1491:12)
    at require (node:internal/modules/helpers:135:16)
    at Object.<anonymous> (/home/runner/work/undici/undici/node-v23.11.0/test/parallel/test-inspector-network-fetch.js:15:16)
    at Module._compile (node:internal/modules/cjs/loader:1734:14) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/runner/work/undici/undici/node-v23.11.0/test/parallel/test-inspector-network-fetch.js'
  ]
}
```

or 

```
✖ SqliteCacheStore works nicely with multiple stores (22.842ms)
✔ SqliteCacheStore maxEntries (4.6677ms)
✔ SqliteCacheStore two writes (0.8061ms)
✔ SqliteCacheStore write & read (1.3409ms)
Error: [Error [ERR_TEST_FAILURE]: failed running after hook] { code: 'ERR_TEST_FAILURE', failureType: 'hookFailed', cause: [Error: EBUSY: resource busy or locked, unlink 'D:\a\undici\undici\cache-interceptor.sqlite'] { errno: -4082, code: 'EBUSY', syscall: 'unlink', path: 'D:\\a\\undici\\undici\\cache-interceptor.sqlite' } }
```

is that something you can help with? I've little to no experience with the CI setup of this repository.

mcollina commented on lib/dispatcher/agent.js:
```diff
@@ -110,6 +110,16 @@ class Agent extends DispatcherBase {
+      if (client.stats) {
+        allClientStats.push([client[kUrl].origin, client.stats])
+      }
+    }
```
instead of an array of arrays, I would use an object.

tdeekens replied:
Sure thing. I assume [this]([52e7d94](https://github.com/nodejs/undici/pull/4157/commits/52e7d940bf3d0669a93b50d5f9fe6954614df57d)) is what you had in mind?

mcollina commented on lib/util/stats.js:
```diff
@@ -0,0 +1,67 @@
+  get size () {
+    return this[kPool][kSize]
+  }
+}
```
I don't really understand this design. Instead of using getters, I would allocate new objects. This ensures that data is stable in time without having to copy them.

tdeekens replied:
To be fair I also didn't but this it the current implementation and I am following it assuming it has reasoning.

tdeekens replied:
👋🏼 What's the implicit agreement here? Refactor as a follow-up, not at all or here?

tdeekens replied:
@metcoder95 what's missing on this PR to land this? Anything expected from my side?

metcoder95 replied:
Just applying the suggested refactor from getters to a plain object (it can also come from a class, but avoiding the getters/setters).

tdeekens replied:
Sure. From what I understand [55566bb](https://github.com/nodejs/undici/pull/4157/commits/55566bbf01230ca947e5a13a3c3da12302876400) is what you had in mind?

tdeekens commented on lib/core/symbols.js:
```diff
@@ -64,6 +64,5 @@ module.exports = {
   kNoProxyAgent: Symbol('no proxy agent'),
   kHttpProxyAgent: Symbol('http proxy agent'),
-  kHttpsProxyAgent: Symbol('https proxy agent'),
-  kStats: Symbol('stats')
```
Not needed anymore. Not assigned as property any longer.

tdeekens commented on lib/dispatcher/client.js:
```diff
@@ -265,7 +262,7 @@ class Client extends DispatcherBase {
 
   get stats () {
-    return this[kStats]
+    return new ClientStats(this)
```
New class every time accessed.

tdeekens commented on lib/util/stats.js:
```diff
@@ -9,58 +9,23 @@ const {
-    this[kClient] = client
-  }
-
-  get connected () {
```
Not getters just instance properties.

tdeekens commented:
@metcoder95 the failing tests appear to be unrelated.

```txt
✔ SqliteCacheStore (16.3615ms)
✖ SqliteCacheStore works nicely with multiple stores (19.9884ms)
✔ SqliteCacheStore maxEntries (4.9747ms)
✔ SqliteCacheStore two writes (0.8221ms)
✔ SqliteCacheStore write & read (1.2954ms)
Error: [Error [ERR_TEST_FAILURE]: failed running after hook] { code: 'ERR_TEST_FAILURE', failureType: 'hookFailed', cause: [Error: EBUSY: resource busy or locked, unlink 'D:\a\undici\undici\cache-interceptor.sqlite'] { errno: -4082, code: 'EBUSY', syscall: 'unlink', path: 'D:\\a\\undici\\undici\\cache-interceptor.sqlite' } }
```

Is this a known issue. Something I have to or you could have a look into?

mcollina reviewed (approved):
lgtm

mcollina commented:
PTAL @ronag @metcoder95

metcoder95 reviewed (approved):
lgtm

tdeekens commented:
@metcoder95 I need some help with the failing tests. I can't see them as relating to the proposed changes. Can you provide some guidance?

metcoder95 commented:
Feel free them, they are not related to your changes as you suggests; I'm starting to dig deeper into it to see if can sort it out.

tdeekens commented:
@metcoder95 sorry to nudge but did you end up anywhere? I tried to understand what's going on but can not find the root cause.

metcoder95 commented:
They are not a blocker for your PR but rather due to other reasons; feel free to ignore them

tdeekens commented:
> They are not a blocker for your PR but rather due to other reasons; feel free to ignore them

Does that mean you would merge the PR or is something else missing to land this?

metcoder95 commented:
LGTM, waiting for @ronag review

tdeekens commented:
Whooop. Thanks all. All approvals are there. I am excited. /cc @metcoder95

tdeekens commented:
Thanks for the support and release. The next version of `@promster/undici` will support exporting agent metrics to Prometheus.
